backend/hidrive/helpers.go

```diff
@@ -0,0 +1,902 @@
+	amount, err = f.updateFileChunked(ctx, path, content, 0, chunkSize, transferLimit)
+
+	if err == nil {
+		info, err = f.resizeFile(ctx, path, amount, modTime)
+	}
+	return amount, info, err
+}
+
+// updateFileChunked updates the content of the existing file at the given path
+// starting at the given offset.
+// Returns the amount of successfully written bytes.
+// If nothing was written this will be 0.
+//
+// Replaces the file contents starting from the given byte offset
+// with the content of the io.Reader.
+// If the offset is beyond the file end, the file is extended up to the offset.
+//
+// The upload is done multiple chunks of the given size in parallel.
+// Therefore this can and be used to upload files of any size efficiently.
+// The number of parallel transfers is limited by transferLimit which should larger than 0.
+//
+// NOTE: Because it is inefficient to set the modification time with every chunk,
+// setting it to a specific value must be done in a separate request
+// after this operation finishes.
+//
+// NOTE: This method uses patchFile and may create sparse files,
+// especially if the upload of a chunk fails unexpectedly.
+// See note about sparse files in patchFile.
+// If any of the uploads fail, the process will be aborted and
+// the first error that occurred will be returned.
+// This is not an atomic operation,
+// therefore if the upload fails the file may be partially modified.
+//
+// This returns fs.ErrorObjectNotFound if the object is not found.
+func (f *Fs) updateFileChunked(ctx context.Context, path string, content io.Reader, offset uint64, chunkSize int, transferLimit int64) (amount uint64, err error) {
+	amount = uint64(0)
+	mutex := new(sync.Mutex)
+	waitgroup, wgCtx := errgroup.WithContext(ctx)
+	transferSemaphore := semaphore.NewWeighted(transferLimit)
+
+	var readErr error
+	startMoreTransfers := true
+	zeroTime := time.Time{}
+	for chunk := uint64(0); startMoreTransfers; chunk++ {
+		// Acquire semaphore to limit number of transfers in parallel.
+		readErr = transferSemaphore.Acquire(wgCtx, 1)
+		if readErr != nil {
+			break
+		}
+
+		// Read a chunk of data.
+		chunkReader, bytesRead, readErr := readerForChunk(content, chunkSize)
+		if bytesRead < chunkSize {
+			startMoreTransfers = false
+		}
+		if readErr != nil || bytesRead <= 0 {
+			break
+		}
+
+		// Transfer the chunk.
+		contentOffset := uint64(chunkSize) * chunk
+		chunkOffset := contentOffset + offset
+		waitgroup.Go(func() error {
+			// After this upload is done,
+			// signal that another transfer can be started.
+			defer transferSemaphore.Release(1)
+			uploadErr := f.patchFile(wgCtx, path, cachedReader(chunkReader), chunkOffset, zeroTime)
+			// Safely update lower-bound
+			// for amount of bytes written before failures.
+			// The errgroup-context wgCtx is canceled implicitly
+			// as soon as the first routine returns an error.
+			if uploadErr == nil {
```

Comment:
This if statement looks a bit untidy...

There is duplicated code and the mutex is being held for longer than necessary.

I dont think the wgCtx checks are needed here

Reply:
I'm resolving this as I'm now reworking the multi-part upload as discussed previously and I'm not sure this will apply anymore.
If you dislike anything with the reworked code please open a new comment when I push the new code. :)